Site Editor Hub: Simplify

youknowriad opened the pull request:
## What and Why?

Currently the site editor renders a "hub" with a site logo, site title and some actions. That hub is persistent and rendered as "fixed" on top of everything.

I think the main reason it's implemented that way is to be able to animate the site logo with go in and out of view mode. That said, I personally believe this is way too complex to accommodate for a small animation.

So this PR is kind of a provocation, how simple can it be if we don't have to implement the animation of the logo. It's not necessarily meant to be merged but at least I want to raise this discussion.

Also, coincidentally, this fixes #60875

Let me know what you think.

youknowriad committed:
Site Editor Hub: Simplify

youknowriad commented on packages/edit-site/src/components/layout/index.js:
```diff
@@ -252,6 +196,11 @@ export default function Layout() {
+										/>
 										<SidebarContent routeKey={ routeKey }>
 											{ areas.sidebar }
 										</SidebarContent>
```
I think we can probably simplify more here in terms of styles and classes.

youknowriad commented on packages/edit-site/src/components/header-edit-mode/index.js:
```diff
@@ -43,6 +46,12 @@ function Header( { setEntitiesSavedStatesCallback } ) {
+				onClick={ () => setCanvasMode( 'view' ) }
+			>
+				<SiteIcon className="edit-site-layout__view-mode-toggle-icon" />
+			</Button>
```
The other good thing here is that this makes the header identical to the post editor and unifying this button is now possible.

annezazu commented:
Quick note that I tried to use WP playground to test this but I can't see the icon when I'm in the canvas to get a better sense of it:

https://github.com/WordPress/gutenberg/assets/26996883/1f5547e0-c5d8-4082-9203-aed53a4b70cf

youknowriad commented:
@annezazu looks like a bug, let me check.

youknowriad committed:
Fix small bug

annezazu commented:
Thank you! Here's a little before/after video to show the differences:

https://github.com/WordPress/gutenberg/assets/26996883/eae2555a-0b8d-4d64-9b21-fe7d0e67962c

Here's the prior PR afaik that animated this: https://github.com/WordPress/gutenberg/pull/48886 I do think some level of magic and polish is lost with this animation gone. Thinking ahead to more personalization efforts with the admin redesign, I imagine this same site "hub" will continue to play a big role to orient folks and some level of animation will go a long way to convey a smoothness and cohesiveness of the experience.

youknowriad commented:
I agree that the animation is nice but what I'm saying is that the way it's implemented today (keeping the same DOM element) is way too complex for what it adds and unless we find a better way to do it, it's not worth it. An example of this is the distraction free bug here https://github.com/WordPress/gutenberg/issues/60875

I think "view transitions" would be perfect here. They already work on Chromium browsers (Chrome, Edge) but not yet in Safari and Firefox. It can be done as a progressive enhancement though.

youknowriad committed:
Animate the hub using view transitions

youknowriad commented:
I pushed a quick prototype for view transitions in Chrome. You can see now in Chrome that the "site icon" is animated. The animation is slightly different than what we have in trunk (I believe we can make it similar to trunk if we really want to) but I'm really excited about this. It's huge improvement over the gymnastic we need to do in our code base to implement animations today.

Also for site hub specifically, it can be acceptable to leverage this as a progressive enhancement: works on Chromium browsers (more than 71% of all users) while giving up on the hub animation for the rest for now and keeping the simplicity of our code base and fixing the distraction free mode bug.

youknowriad committed:
Hide original icon when going into full editor

youknowriad committed:
Fix sidebar padding

youknowriad committed:
Fix border radius

youknowriad committed:
No fadein/fadeout animation for the frame

youknowriad committed:
Delay the site icon animation

youknowriad committed:
Fix border radius

annezazu commented:
Tested again and it doesn't feel very smooth or polished compared to before:

https://github.com/WordPress/gutenberg/assets/26996883/f1194f86-d790-414b-a7cf-e62a9847faaa

This is especially the case when returning to the overall navigation where the site icon zooms in almost too much. When entering the canvas, it feels delayed. I'm not sure if there's a way to make it more seamless while still getting the benefits of this PR as I definitely want to see this bug resolved https://github.com/WordPress/gutenberg/issues/60875. 

As is though, this PR isn't quite there.

youknowriad committed:
Delay only one direction of the animation

youknowriad commented:
I updated this PR to delay only one side of the animation. I think this is way better. Thoughts?

draganescu reviewed (commented):
Distraction Free Mode is still broken with this in Firefox:


https://github.com/WordPress/gutenberg/assets/107534/ac288f2f-ab6c-49e4-84d1-caf58eea7b17

The animation works very well in other browsers.

youknowriad commented:
@draganescu  what's broken exactly in Firefox? For me it's working, so I wonder if we're talking about a different unrelated bug.

draganescu commented:
@youknowriad as the video shows in Firefox when the edit mode is entered and DFM is on the  interface header contains no buttons and stays there empty.

On this PR in other browsers the current bug in trunk where the header does not appear at all in DFM is fixed. So the fix is incomplete somehow in FF and this PR introduces the problem in FF.

youknowriad commented:
@draganescu I can reproduce that bug in any browser, it depends on where you click to enter "edit mode". I suspect that it's another bug unrelated with the current fix but I'll take a look anyway.

youknowriad commented:
Ok, so the bug is due to multiple header animations conflicting with each other, I solved it by simplifying and unifying the header animation in a single place.

youknowriad committed:
Fix linting issue

jeryj reviewed (commented):
We'll need a way to manage focus between the site hub and the header. On `trunk`, the button for each is the same, so we don't need to do anything to move focus. On this, we're removing the clicked button and it doesn't look like we're placing focus anywhere.

youknowriad commented:
@jeryj I'm guessing that you're thinking about the move from "edit" to "view" mode right? because for the opposite direction, it's probably the same as trunk.

jeryj commented:
Yes, edit to view mode. I forgot that going from view to edit, you enter by selecting the canvas, which places focus on the canvas correctly.

youknowriad committed:
Handle focus properly when leaving edit mode

youknowriad commented:
Added focus handling when leaving edit mode.

draganescu reviewed (commented):
I am unsure how to deal with this PR. One one hand the PR fixes the bug in `trunk` which makes the DFM impossible to exit in the site editor. On the other hand the UI of DFM is severely changed in terms of animation:

- there is a noticeably bigger delay on the hover effect
- the animation is just a simple fade, a big contrast to the rest of the editor UI

I think it's a better overall experience offered by this PR, given that in chrome the hub animation is nearly identical to `trunk` and the other browsers show some minute differences. But scoped only to DFM it's worse.

As a note, I remember Matias was pretty precise and persistent on how the hover animation of DFM. I wasn't even able to land a change that would use CSS transitions to clean up some of the ~~nesting hell~~ complexity introduced by motion components. The animation had to:

- be timed correctly
- be orchestrated correctly
- have the background fade in
- have the hub slide from the left, the left tools from the top and the right tools from the right

... given the bug in `trunk` maybe this PR is OK and should land and then strive to fix the UI for DFM during the WP beta cycle? cc @jasmussen

jasmussen commented:
> ... given the bug in trunk maybe this PR is OK and should land and then strive to fix the UI for DFM during the WP beta cycle?

Yes, that sounds good to me.

youknowriad commented:
> should land and then strive to fix the UI for DFM during the WP beta 

@draganescu Can you explain more what are the remaining DFM issues in this PR?

draganescu commented:
> Can you explain more what are the remaining DFM issues in this PR?

I thought I did, extensively, but here is some video 😁 

WP 6.5.4-alpha, no Gutenberg:

https://github.com/WordPress/gutenberg/assets/107534/164ed7e4-6765-4326-9b71-e2fe1c03c74a

WP 6.5.4-alpha, with Gutenberg on this PR:


https://github.com/WordPress/gutenberg/assets/107534/7ccc237f-c072-4588-9a5b-d2e6f2de850b



Differences:

- the hub does not slide from the left
- the tools don't slide from the top
- it all fades
- there is an increased delay hovering the header until the animation kicks in

youknowriad commented:
Thanks for sharing the details. I agree they are different but my opinion is that the animation in this PR is better. Obviously it's very arguable and subjective. What is objective is that in terms of implementation and code base, it's way simpler (so better).

For the delays, I actually did it on purpose to delay the hover effect but I didn't delay it on leave (which I can do if needed). I'm happy to tweak the delays.

Would love some design opinion on the DFM animation.

youknowriad committed:
Fix hub post rebase

jasmussen commented:
I don't think the difference in animation is a meaningful differential to block this PR. Though I do agree it can be improved, I'd argue that's more about animating the whole top-bar in from the top, rather than fading it in.

youknowriad commented:
I don't want to argue too much about the animation, so to unblock this PR (and another PR :P ) I extracted some of the animation changes into this smaller PR https://github.com/WordPress/gutenberg/pull/62167 

Let's try to land this related PR first. It touches the header animation but hopefully try to stay as close as possible to the trunk animation. Would love reviews there.

youknowriad commented:
Now that #62167 landed, I've rebased this PR. So now the DFM bug if fixed in the site editor and the animation is correct as well.

richtabor commented:
This tests well (especially in Chromium browsers).

richtabor commented on packages/edit-site/src/components/layout/style.scss:
```diff
@@ -158,17 +127,36 @@
+/* stylelint-disable -- Disable reason: View Transitions not supported properly by stylelint. */
+html:active-view-transition-type(canvas-mode-edit) {
+	&::view-transition-group(toggle) {
+		animation-delay: 250ms;
```
It's tiny, but `225ms` gives the frame enough time to move over before scaling back the site logo.

richtabor commented on packages/edit-site/src/components/site-hub/index.js:
```diff
@@ -56,141 +43,63 @@ const SiteHub = memo( ( { isTransparent, className } ) => {
+							label={ __( 'Go to the Dashboard' ) }
+							className="edit-site-layout__view-mode-toggle"
+							style={ {
+								transform: 'scale(0.5)',
```
Perhaps it's ok, but there is a bit of a difference in site icon size from trunk: 

![CleanShot 2024-05-31 at 08 12 50](https://github.com/WordPress/gutenberg/assets/1813435/b87874c7-51ca-4d3c-8c7f-06ca2fc61304)

Simply increasing the scale doesnt work great though, as then the WP SVG also gets increased slightly.

richtabor commented:
It's close. There's still a bit of a flash/odd view as the editor header is moving out, when you're exiting edit mode. 

https://github.com/WordPress/gutenberg/assets/1813435/683f43f5-c45d-488e-be16-f5bfab7aa48e

https://github.com/WordPress/gutenberg/assets/1813435/78c6a2bf-10f6-4f5d-aebd-101870f70903

youknowriad committed:
Better delay

youknowriad commented:
- Rebased the PR. I'm not able to reproduce the toolbars being visible anymore (but let me know if I'm not testing properly).
- I adjusted the delay like above.
- I'm failing to understand the flash of the W svg: Any help there would be appreciated.

richtabor commented:
> I'm failing to understand the flash of the W svg: Any help there would be appreciated.

I shortened the clip to see it here: 

https://github.com/WordPress/gutenberg/assets/1813435/2fca6556-787d-412b-80ff-99359093a427

youknowriad committed:
Fix lingering block toolbar

youknowriad commented:
@draganescu I fixed the lingering block toolbar.

youknowriad committed:
Simplify

youknowriad commented:
@richtabor I think I understand the "W" animation issue a little bit better. The issue comes from the fact that when using the "w" logo or a random site icon, the proportions of the "w" logo are not the same inside and outside the frame. Basically when we're in the editor, we add a padding around the icon but this padding don't exist in view mode which means the two icons actually don't match between edit and view mode, so when we move between one and the other, there's a crossfade animation that is slightly visible because of that padding.

I'm not sure how to solve it without removing that padding (which increases the sizes of the icon in the editor) or adding it to both view and edit mode (which make the icon smaller in view mode).

For sites with an actual image, the proportions are the same.

draganescu commented on packages/edit-site/src/store/private-actions.js:
```diff
@@ -13,39 +13,56 @@ import { store as editorStore } from '@wordpress/editor';
+		const switchCanvasMode = () => {
+			const isMediumOrBigger =
+				window.matchMedia( '(min-width: 782px)' ).matches;
+			registry.dispatch( blockEditorStore ).clearSelectedBlock();
```
When top toolbar is ON and we exit edit mode to view mode there is a short moment when the header is removing the top toolbar and _then_ the animation moves forward. Can we do it in an unblocking way? Should it be already unblocking? 🤷🏻

youknowriad replied:
What do you mean "unblocking way"?

youknowriad replied:
I'm not sure that's entirely what you meant but I noticed that we're triggering a big number of actions when switching canvas mode, so I batched everything in 6c58262cf3

I think the animation is smoother with that.

draganescu replied:
Yes this is exactly what I meant :) The animation doesn't look like it's waiting on something to be over before starting, it just goes now. 👏🏻

draganescu reviewed (approved):
The ghost toolbars are fixed. Left a small problem which I consider a nit.

From my perspective this PR is a good step forward in terms of improving the code base. Given it preserves the way the UI behaved in terms of both interaction and feedback I think this can land given the tests pass. @richtabor do you think the W issue is blocking?

youknowriad committed:
Avoid the types view transitions as it's not supported in all browsers that support view transitions

youknowriad committed:
Fix linting

youknowriad committed:
Batch updates when switching editor modes

youknowriad commented:
Would love another opinion here @richtabor @jasmussen or @mtias before merging this one.

youknowriad committed:
Render the back button conditionally

youknowriad committed:
Fix e2e tests

youknowriad committed:
More stable e2e tests

stokesman commented:
I'm here to invite anyone still subscribed to have a look at #62386 which follows up on this by restoring the previous transition. To me, the transition after this is too degraded to overlook. I tried various ways of fixing it and that’s what seems best to me.

ellatrix commented:
This was cherry-picked to the wp/6.6 branch.
